docs/timezone-draft.md

```diff
@@ -0,0 +1,160 @@
+
+These functions gain an `options` parameter with an `idToTimeZone` option, whose value is a function returning the appropriate time zone object for the `id` passed in, or `null` to indicate that the time zone doesn't exist.
+
+If the option is not given, `Temporal.TimeZone.idToTimeZone()` is called.
+This is a new static method of `Temporal.TimeZone`.
+
+> **FIXME:** Note that there is an issue open regarding this behaviour for calendars: [#294](https://github.com/tc39/proposal-temporal/issues/294)
+
+```javascript
+function idToTimeZone(id) {
+  if (id === 'Mars/Olympus_Mons')
+    return new MarsTime();
+  return Temporal.TimeZone.idToTimeZone(id);
+}
+
+Temporal.Absolute.from('2020-04-01T18:16+00:00[Mars/Olympus_Mons]', { idToTimeZone })
+  // returns the Absolute corresponding to 2020-04-01T18:16Z,
+  // assuming that MarsTime instances have an offset of +00:00 at that instant
+
+Temporal.TimeZone.from('2020-04-01T18:16+00:00[Mars/Olympus_Mons]', { idToTimeZone })
+  // returns a new MarsTime instance
+```
+
+> **FIXME:** Passing a custom time zone's identifier to the built-in `Temporal.TimeZone` constructor currently doesn't work: `new Temporal.TimeZone('Mars/Olympus_Mons')` throws.
+> However, this must change, because implementations would need to call `super(id)` to set the _[[Identifier]]_ and _[[InitializedTemporalTimeZone]]_ internal slots.
+> Maybe we need to make `Temporal.TimeZone` not directly constructable?
+
+In order to perform attenuation, an implementation would monkeypatch the `Temporal.TimeZone.idToTimeZone()` function which performs the built-in mapping, and change the list of allowed time zones that `Temporal.TimeZone` iterates through.
+Or it could just replace the `Temporal.TimeZone` class altogether:
+
+```javascript
+// For example, to allow only offset time zones:
+
+TemporalTimeZone_idToTimeZone = Temporal.TimeZone.idToTimeZone;
+Temporal.TimeZone.idToTimeZone = function (id) {
+  if (/^[+-]\d{2}:?\d{2}$/.test(id))
+    return TemporalTimeZone_idToTimeZone(id);
+  return null;
+}
+Temporal.TimeZone[Symbol.iterator] = function* () { return null; }
+
+// or, to replace the built-in implementation altogether:
+
+Temporal.TimeZone = FrozenTimeZoneImplementation;
+```
+
+## Implementation of a custom time zone
+
+### TimeZone interface
+
+```javascript
+class Temporal.TimeZone {
+  /** Sets the [[Identifier]] internal slot to @id, and creates the
+   * [[InitializedTemporalTimeZone]] internal slot. A subclassed custom
+   * time zone must chain up to this constructor. */
+  constructor(id : string) : Temporal.TimeZone;
+
+  // Methods that a subclassed custom time zone must implement
+
+  /** Given an absolute instant returns this time zone's corresponding
+   * UTC offset, in nanoseconds (signed). */
+  getOffsetAtInstant(absolute : Temporal.Absolute) : number;
+
+  /** Given the calendar/wall-clock time, returns an array of 0, 1, or
+   * 2 absolute instants that are possible points on the timeline
```

Comment:
> Do we want to enforce the maximum of two elements?

I don't know, I was wondering if it would theoretically be possible for a time zone to decide to have two DST transitions back-to-back so that a DateTime was repeated three times. That wouldn't be handled by getAbsoluteFor, since the disambiguation parameter only takes "earlier" or "later" into account, but possibleInstants could be used by user code to deal with that case.

If we did want to enforce the maximum then it might make more sense to return an object `{ earlier: time1, later: time2 }` and just have getAbsoluteFor take the key that corresponds to the disambiguation parameter, or throw if the two are different in "reject" mode. That would move the responsibility for deciding what to do in the case of 0 absolutes found, from getAbsoluteFor into possibleInstants, and therefore into user code in the case of custom time zones.

> Do we want to enforce this returns an array, rather than an iterable?

In the current form I'm not sure it makes sense to return an iterable, since the first thing we do is check its length.

> Should this comment say something about how we'd pick one of them?

I'll add this.